src/hotspot/share/gc/shenandoah/shenandoahFullGC.cpp

```diff
@@ -1153,17 +1153,17 @@ ShenandoahGenerationalHeap::TransferResult ShenandoahFullGC::phase5_epilog() {
     }
 
     heap->collection_set()->clear();
-    size_t young_cset_regions, old_cset_regions;
-    size_t first_old, last_old, num_old;
-    heap->free_set()->prepare_to_rebuild(young_cset_regions, old_cset_regions, first_old, last_old, num_old);
-
+    size_t young_cset_regions, old_cset_regions, first_old, last_old, num_old;
+    ShenandoahFreeSet* free_set = heap->free_set();
+    ShenandoahRebuildLocker rebuild_locker(free_set->lock());
```

Comment:
Should you not create a scope around lines 1158 to line 1167, since you don't want to hold the rebuild lock as soon as the rebuild is done (i.e. immediately following `finish_rebuild()`)?

Reply:
I'll tighten up the context for the rebuild lock.  I was thinking that set_mark_incomplete() and clear_cancelled_gc() would be "fast enough" that it wouldn't matter to hold the rebuild_lock this much longer, but I agree it is better to release the lock as soon as possible.